File: easy/0653_two_sum_iv_input_is_a_bst/solution.cpp
#include "../../utils/cpp/nodes.hpp"
#include <stack>
#include <vector>

class Solution {
public:
  bool findTarget(utils::TreeNode *root, int k) {
    utils::TreeNode *curr = root;

    std::stack<utils::TreeNode *> s;
    std::vector<int> sortedVals;

    while (curr || !s.empty()) {
      while (curr) {
        s.push(curr);
        curr = curr->left;
      }

      sortedVals.push_back(s.top()->val);
      curr = s.top()->right;

      s.pop();
    }

    size_t left = 0;
    size_t right = sortedVals.size() - 1;

    while (left < right) {
      int currSum = sortedVals[left] + sortedVals[right];

      if (currSum == k)
        return true;
      else if (currSum < k)
        ++left;
      else
        --right;
    }

    return false;
  }
};

File: easy/0653_two_sum_iv_input_is_a_bst/Solution.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
import nodes.TreeNode;

class Solution {
  public boolean findTarget(TreeNode root, int k) {
    TreeNode curr = root;

    final Deque<TreeNode> stack = new ArrayDeque<>();
    final List<Integer> sortedVals = new ArrayList<>();

    while (curr != null || !stack.isEmpty()) {
      while (curr != null) {
        stack.push(curr);
        curr = curr.left;
      }

      final TreeNode node = stack.pop();

      sortedVals.add(node.val);
      curr = node.right;
    }

    int left = 0;
    int right = sortedVals.size() - 1;

    while (left < right) {
      final int currSum = sortedVals.get(left) + sortedVals.get(right);

      if (currSum == k) {
        return true;
      } else if (currSum < k) {
        ++left;
      } else {
        --right;
      }
    }

    return false;
  }
}
